File: 백준/Gold/2668. 숫자고르기/숫자고르기.cc
#include <bits/stdc++.h>
using namespace std;
int n;
int arr[101];
bool vis[101];
bool finished[101];
vector<int> result;

void dfs(int cur, int start) {
    vis[cur] = true;
    int next = arr[cur];
    
    if(!vis[next]) {
        dfs(next, start);
    } else if(next == start) {
         result.push_back(start);   
    }  
    finished[cur] = true;
}

int main() {
    cin >> n;
    for(int i=1; i<=n; i++) {
        cin >> arr[i];
    }
    for(int i=1; i<=n; i++) {
        fill(vis, vis+n+1, false);
        dfs(i, i);
    }
    sort(result.begin(), result.end());

    cout << result.size() << "\n";
    for(int i=0; i<result.size(); i++) {
        cout << result[i] << "\n";
    }
}

File: 백준/Gold/2668. 숫자고르기/숫자고르기.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static int n;
    static int[] arr;
    static boolean[] vis;
    static boolean[] finished;
    static List<Integer> ans = new ArrayList<>();
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());

        arr = new int[n+1];
        vis = new boolean[n+1];
        finished = new boolean[n+1];

        for(int i=1; i<=n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        for(int i=1; i<=n; i++) {
            if(!vis[i]) {
                dfs(i);
            }
        }

        Collections.sort(ans);
        System.out.println(ans.size());
        for(int x : ans) System.out.println(x);
    }

    static void dfs(int cur) {
        vis[cur] = true;
        int next = arr[cur];

        if(!vis[next]) {
            dfs(next);
        } else if (!finished[next]) {
            while(next != cur) {
                ans.add(next);
                next = arr[next];
            }
            ans.add(cur);
        }
        finished[cur] = true;
    }
}